File: src/main/java/com/freya02/botcommands/api/ExceptionHandler.java
package com.freya02.botcommands.api;

import net.dv8tion.jda.api.events.Event;
import net.dv8tion.jda.api.hooks.EventListener;

/**
 * Similar to {@link EventListener}
 */
public interface ExceptionHandler {
	/**
	 * <b>Note: You are sent a generic Event, you will need to check it against SlashCommandInteractionEvent, GuildMessageReceivedEvent, etc... in order to differentiate events</b>
	 *
	 * @param context The current context
	 * @param event The event which triggered this exception
	 * @param throwable The throwable which got threw
	 */
	void onException(BContext context, Event event, Throwable throwable);
}

File: src/main/kotlin/com/freya02/botcommands/internal/ExceptionHandler.kt
package com.freya02.botcommands.internal

import com.freya02.botcommands.api.BContext
import com.freya02.botcommands.internal.utils.unwrap
import mu.KLogger
import net.dv8tion.jda.api.events.Event

internal class ExceptionHandler(private val context: BContext, private val logger: KLogger) {
    fun handleException(event: Event, e: Throwable, locationDescription: String) {
        val unreflectedException = e.unwrap()
        val handler = context.globalExceptionHandler
        if (handler != null) {
            handler.onException(context, event, unreflectedException)
            return
        }

        val errorMessage = "Uncaught exception in $locationDescription"
        logger.error(errorMessage, unreflectedException)
        context.dispatchException(errorMessage, unreflectedException)
    }
}